### PR TITLE
[stable/prometheus-operator] Add support for creating a Service and Ingress for each Prometheus replica

### PR DESCRIPTION
**Pull request by shaunmansell**
No

##### What
Adding support for creating separate Service and Ingress resources that will map
to the standard web port of each Prometheus server replica.

We're using the `prometheus.prometheusSpec.replicas` value to determine how many
of each resource type to be created, and to vary resource names and values.

Because we're using a single set of values to control certain properties, we have to
drop support for setting parameters that need to be unique within a each resource.
Primarily this means no IP configuration options for the Service resource.

The ingress resources are created with a single host entry that is compiled by taking
two values (`hostPrefix` and `hostDomain`). The final hostname format is
`hostPrefix`-`replicaIndex`.`hostDomain`, where `replicaIndex` will be 0, 1, 2 etc.

##### Why
My use case for this is to expose each Prometheus server directly via an ingress definition so I can:
- Support Grafana datasources targeting each replica independently. We currently do this via the `prometheus-operated` headless service, however I want to get further metrics on the performance of the inbound requests that Prometheus does not current record internally.
- Scrape each Prometheus server `/metrics` from outside the Kubernetes cluster the server is running in
- Capture requests statistics via the ingress controller metrics to monitor Prometheus API request success and latency

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/chart]`)

### CODE REVIEW

**Comment by shaunmansell** [comment]
I was also considering a PR for Alertmanager to support creating service and ingress resources for each replica, where that is primarily to support routing Prometheus communication through as proxy service that allows us to track inbound requests and their outcome and processing time.

I'll await feedback on this PR before doing any work on the Alertmanager side.

**Comment by vsliouniaev** [comment]
I think this is useful from a federation point of view
/ok-to-test

File: stable/prometheus-operator/templates/prometheus/serviceperreplica.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
+{{- if and .Values.prometheus.enabled .Values.prometheus.servicePerReplica.enabled }}
+{{- $count := .Values.prometheus.prometheusSpec.replicas | int -}}
+{{- $serviceValues := .Values.prometheus.servicePerReplica -}}
+{{ range $i, $e := until $count }}
+---
```
Comment (vsliouniaev) [inline stable/prometheus-operator/templates/prometheus/serviceperreplica.yaml]: This should be a list, rather than individual resources

**Review by vsliouniaev — CHANGES_REQUESTED** [review]

File: stable/prometheus-operator/templates/prometheus/serviceperreplica.yaml (hunk 1 above)
Comment (shaunmansell) [reply stable/prometheus-operator/templates/prometheus/serviceperreplica.yaml]: I've updated both the service and ingress resources to be `ServiceList` and `IngressList` kinds as I took that to be what you meant.

For my own interest, is there any particular reason to prefer these over a templated list of separate resources like what I had originally?

**Comment by shaunmansell** [comment]
@vsliouniaev thanks for taking a look at this!

I've updated the service and ingress resources as you noted and bumped the chart version to resolve the previous conflict. Let me know if there's anything else needed.

**Pull request merged**